File: Data Structures & Algorithms/min-cost-climbing-stairs/submission-2.py
class Solution:
    def minCostClimbingStairs(self, cost: List[int]) -> int:
        n = len(cost)

        memo = {}
        def dfs(i):
            if i >= n:
                return 0
            
            if i in memo:
                return memo[i]
            
            single_step = cost[i] + dfs(i + 1)
            two_step = cost[i] + dfs(i + 2)

            memo[i] = min(single_step, two_step)
            return memo[i]
        
        return min(dfs(0), dfs(1))

File: Data Structures & Algorithms/coin-change/submission-2.cpp
class Solution {
public:
    unordered_map<int, int> memo;
    int dfs(vector<int>& coins, int amount){
        if (amount == 0) return 0;
        if (memo.find(amount) != memo.end()) {
            return memo[amount];
        }
        int res = 1e9;
        for (int coin: coins) {
            if (amount - coin >= 0){
                int result = dfs(coins, amount - coin);
                if (result != INT_MAX) {
                    res = min(res, 1 + result);
                }
                
            }
        }
        memo[amount] = res;
        return res;
    }



    int coinChange(vector<int>& coins, int amount) {
        int minCoins = dfs(coins, amount);
        return (minCoins >= 1e9) ? -1 : minCoins;
    }
};
